service/history/replication/task_ack_manager.go

```diff
@@ -152,10 +176,54 @@ func (t *TaskAckManager) GetTasks(ctx context.Context, pollingCluster string, la
 		t.logger.Error("error updating replication level for hydrated task store", tag.Error(err), tag.OperationFailed)
 	}
 
-	t.logger.Debug("Get replication tasks", tag.SourceCluster(pollingCluster), tag.ShardReplicationAck(lastReadTaskID), tag.ReadLevel(readLevel))
-	return &types.ReplicationMessages{
-		ReplicationTasks:       replicationTasks,
-		HasMore:                hasMore,
-		LastRetrievedMessageID: readLevel,
-	}, nil
+	t.logger.Debug(
+		"Get replication tasks",
+		tag.SourceCluster(pollingCluster),
+		tag.ShardReplicationAck(lastReadTaskID),
+		tag.ReadLevel(msgs.LastRetrievedMessageID),
+	)
+	return msgs, nil
+}
+
+// shrinkMessagesBySize shrinks the replication messages by removing the last replication task until the total size is allowed
+func (t *TaskAckManager) shrinkMessagesBySize(msgs *types.ReplicationMessages) error {
+	// if there are no replication tasks, do nothing
+	if len(msgs.ReplicationTasks) == 0 {
+		return nil
+	}
+
+	maxSize := t.maxReplicationMessagesSize
+
+	for {
+		totalSize := t.replicationMessagesSizeFn(msgs)
+
+		// if the total size is allowed, return the replication messages
+		if totalSize < maxSize {
+			return nil
+		}
+
+		lastTask := msgs.ReplicationTasks[len(msgs.ReplicationTasks)-1]
+		t.logger.Warn("Replication messages size is too large. Shrinking the messages by removing the last replication task",
```

Comment:
my suspicion is that this will be pretty noisy if we're logging once per message, but I guess we can experiment and see. It seems likely we'll hit the limit in bursts quite a bit.

Reply:
might be worth throwing in a metric indicating there's trimming going on

Reply:
I think the number of new logs will be similar to the number of `Replication messages did not fit in the response (history host)`, but the occurrence of the second will be much less, so let's see how it goes :) 

We already have [this metric](https://github.com/cadence-workflow/cadence/pull/6719/files#diff-2fcd79f6708c116eb367f4d4671005b768ed469aa4f5575fdb499d020d20b36dR169). It doesn't show that it caused by the shrinking, however I don't see other reasons of having this diff.